### ISSUE TITLE
Fork of k/k to be used for Release Engineering integration tests

### ISSUE THREAD

**Issue by justaugustus** [kubernetes/org#1474]
### New Repo, Staging Repo, or migrate existing
New repo, fork of k/k

### Requested name for new repository
`kubernetes/release-test`

### Which Organization should it reside
kubernetes (It's helpful here to keep it in the kubernetes org, since all of the SIG Release assignment/ping teams are there.)

### If not a staging repo, who should have admin access
https://github.com/kubernetes/org/blob/d39fbb43a0c7637a9bb5bbedcd66a647f4aaada3/config/kubernetes/sig-release/teams.yaml#L364-L370

### If not a staging repo, who should have write access
https://github.com/kubernetes/org/blob/d39fbb43a0c7637a9bb5bbedcd66a647f4aaada3/config/kubernetes/sig-release/teams.yaml#L201-L219

### If not a staging repo, who should be listed as approvers in OWNERS
Already configured.

### If not a staging repo, who should be listed in SECURITY_CONTACTS
Already configured.

### What should the repo description be
Release Engineering fork of kubernetes/kubernetes

### What SIG and subproject does this fall under in sigs.yaml
This is part of the Release Engineering subproject for SIG Release

### Approvals
Discussed on Release Engineering calls, but I'll put this to something official prior to creation.

Authoritative requirements are here: https://git.k8s.io/community/github-management/kubernetes-repositories.md

tl;dr (but really you should read the linked doc, this may be stale)
- If this is a core repository, then sig-architecture must approve
- If this is a SIG repository, then this must follow the procedure spelled out
  in that SIG's charter

### Additional context for request
The Release Engineering subproject uses various tools to manipulate the content in the kubernetes/kubernetes repo.

As we work to [refactor existing bash scripts into go tools](https://github.com/kubernetes/release/issues/918), we need a playground that we can use, reset, destroy to target integration testing against.

@kubernetes/release-engineering @kubernetes/owners @kubernetes/sig-architecture-feature-requests

/milestone v1.18
/area release-eng
/sig release
/priority important-soon
/kind feature

**Comment by tpepper** [comment]
I suspect this might be more work, but...if we're going to change something:

I'd like to see us consider change tooling to explicitly reference a (repo,branch) tuple.  If k/k:master is today implicit, or only the k/k part as a repo is configurable (ie: "kubernetes/kubernetes" is in a bunch of repo variables in test-info yaml...and the proposal is to reference instead some new k/otherk, can we make existing k/k:master explicit and configurable?  Then we could have some references to k/k:releng-test in addition to k/k:master.  Maybe that's even already possible today, but we're just accustomed to only stating and changing the repo?  Or it's a fundamental design point everywhere in the test-infra code and near impossible to change at this point to reference a branch?

**Comment by tpepper** [comment]
We do already have sig-release tests that are branch specific, with repo explicit and branch explicit, eg:
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-release/release-branch-jobs/1.14.yaml#L370
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-release/release-branch-jobs/1.14.yaml#L373

**Comment by saschagrunert** [comment]
I like this idea and see two scenarios which we could cover with such a repository, for which I have a question:

- **Having a playground for developing krel.**
  - Do we have a chance to mirror the repos with all its branches and tags?
- **Integration testing of krel / release notes.**
  - How can we maintain a reproducible starting state for writing the integration/e2e tests?
  (Thinking: We could create synthetic branches in the fork which are just used to isolate single test cases)

**Comment by justaugustus** [comment]
@saschagrunert is correctly describing the scenarios I have in mind.

**Comment by justaugustus** [comment]
/remove-priority important-soon
/priority critical-urgent
/assign @cblecker @nikhita @mrbobbytables @spiffxp 

https://github.com/kubernetes/kubernetes/issues/86182 makes a decent case for us having a playground to vet release changes against.

cc: @liggitt @ixdy

**Comment by neolit123** [comment]
i'm surprised this proposal for a new repository is not approved yet.

my suggestion for the name is `kubernetes/release-test`.

**Comment by justaugustus** [comment]
> my suggestion for the name is `kubernetes/release-test`.

I like this, @neolit123. Updated the description with this suggestion.

**Comment by cblecker** [comment]
Two questions:
- How will you keep this in sync with kubernetes/kubernetes? Release team's own tools?
- Have you taken this to sig-arch for approval?

**Comment by neolit123** [comment]
i personally think there is not need to auto-sync this repository with k/k. it is a test bed for building a release, test tagging and so on. manual syncs can be done with PRs.

**Comment by tpepper** [comment]
As of the [kubeadm move out of k/k KEP](https://github.com/kubernetes/enhancements/pull/1425), if we go the route proposed in this issue, we will need not just a fork of k/k but also a fork of the new k/kubeadm.

**Comment by tpepper** [comment]
We probably don't want these repos to auto sync.  I foresee them operating as non-fast-forward repos/branches with us moving HEAD forward and backwards, tags being added and deleted.  To experiment fully we'd want to repeatedly run different k/release tools against the same repo state, but each run will modify that state.

**Comment by neolit123** [comment]
> As of the kubeadm move out of k/k KEP, if we go the route proposed in this issue, we will need not just a fork of k/k but also a fork of the new k/kubeadm.

having multiple test repositories seems fine, but if this sees objection then the release tooling has to consider using private forks.

**Comment by justaugustus** [comment]
> having multiple test repositories seems fine, but if this sees objection then the release tooling has to consider using private forks.

Hang tight on making decisions for us.

**Comment by justaugustus** [comment]
@cblecker --

> * How will you keep this in sync with kubernetes/kubernetes? Release team's own tools?

No sync is required at the moment. I think eventually we may have some tool that resets the branches every day or something, but right now, we just need something to test against that I can grant other Releng staff access to.

> * Have you taken this to sig-arch for approval?

Will shoot a note to the list.

**Comment by neolit123** [comment]
> Hang tight on making decisions for us.

i don't think i'm making decisions for you.

**Comment by spiffxp** [comment]
I feel like it's worth considering a separate org as a bright line to avoid confusion.  Especially if we're talking multiple test repos.  That said, repos in the kubernetes org are ultimately sig arch's call to make, if they're fine with it, we're fine with it.

**Comment by nikhita** [comment]
Removing urgent priority since this has been open for a while. Please add it again if needed!

/remove-priority critical-urgent

**Comment by saschagrunert** [comment]
I'm now wondering if we still need it or could mock-test around it.

**Comment by saschagrunert** [comment]
I’d like to consider this as not resolvable for now. We could think about having a stage+release (mock) postsubmit Job but this will take 1h per CI run. 
/close
